### PR TITLE
Update to coverage-04.dtd and support LineRate, line-valid and line-covered

### PR DESCRIPTION
**Pull request by yaoyaozong**
1. Updated to coverage-04.dtd format
2. Add the calculation for linerate, line-valid and line-covered

After these changes, the converted xml can be loaded correctly in Visual Studio Team Services(vsts) and show the correct results in vsts Test coverage page.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,13 +5,15 @@ This is a simple helper tool for generating XML output in [Cobertura](http://cob
...
+UPDATES: enhanced AlekSi's tool to make it support cobertura coverage-04.dtd format. And make it show correct coverage data on vsts 
+
 Installation
 ------------
 
 Just type the following to install the program and its dependencies:
 
     $ go get github.com/axw/gocov/...
-    $ go get github.com/AlekSi/gocov-xml
+    $ go get github.com/yaoyaozong/gocov-xml
```
Comment (AlekSi) [inline README.md]: Could you please revert this change?

File: README.md [hunk 2]
```diff
@@ -5,13 +5,15 @@ This is a simple helper tool for generating XML output in [Cobertura](http://cob
 for CIs like [Jenkins](https://wiki.jenkins-ci.org/display/JENKINS/Cobertura+Plugin) and others
 from [github.com/axw/gocov](https://github.com/axw/gocov) output.
 
+UPDATES: enhanced AlekSi's tool to make it support cobertura coverage-04.dtd format. And make it show correct coverage data on vsts
```
Comment (AlekSi) [inline README.md]: Please remove this.

File: README.md [hunk 3]
```diff
@@ -25,3 +27,6 @@ Authors
 
 * [Alexey Palazhchenko (AlekSi)](https://github.com/AlekSi)
 * [Yukinari Toyota (t-yuki)](https://github.com/t-yuki)
+* [Marin Bek (marinbek)](https://github.com/marinbek)
+* [Alex Castle (acastle)](https://github.com/acastle)
+* [Billy Yao (yaoyaozong)](https://github.com/yaoyaozong)
```
Comment (AlekSi) [inline README.md]: 👍

File: coverage-04.dtd [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  <!ATTLIST line hits   CDATA #REQUIRED>
+  <!ATTLIST line branch CDATA "false">
+  <!ATTLIST line condition-coverage CDATA "100%">
+
+  <!ELEMENT conditions (condition*)>
+
+  <!ELEMENT condition EMPTY>
+  <!ATTLIST condition number CDATA #REQUIRED>
+  <!ATTLIST condition type CDATA #REQUIRED>
+  <!ATTLIST condition coverage CDATA #REQUIRED>
```
Comment (AlekSi) [inline coverage-04.dtd]: Can you please add a source URL to that file to README? So I and everyone else can find it easily.

File: gocov-xml.go [hunk 5]
```diff
@@ -55,6 +67,7 @@ type Line struct {
 
 func main() {
 	var r struct{ Packages []gocov.Package }
+	var total_lines, total_hits int64 = 0, 0
```
Comment (AlekSi) [inline gocov-xml.go]: That should be
```go
var totalLines, totalHits int64
```

No snake_case. Initialization with zeros is implicit.

File: gocov-xml.go [hunk 6]
```diff
@@ -109,30 +122,43 @@ func main() {
 
 			// convert statements to lines
 			lines := make([]Line, len(gFunction.Statements))
+			func_hits := 0
```
Comment (AlekSi) [inline gocov-xml.go]:
```go
var funcHits int
```

File: gocov-xml.go [hunk 7]
```diff
@@ -109,30 +122,43 @@ func main() {
...
 			for i, s := range gFunction.Statements {
 				lineno := tokenFile.Line(tokenFile.Pos(s.Start))
 				line := Line{Number: lineno, Hits: s.Reached}
+				if int(s.Reached) > 0 {
+					func_hits += 1
+				}
 				lines[i] = line
 				class.Lines = append(class.Lines, line)
 			}
+			line_rate := float32(func_hits) / float32(len(gFunction.Statements))
```
Comment (AlekSi) [inline gocov-xml.go]: `lineRate`

File: gocov-xml.go [hunk 8]
```diff
@@ -80,7 +93,7 @@ func main() {
 			className, methodName := s[len(s)-2], s[len(s)-1]
 			class := classes[className]
 			if class == nil {
-				class = &Class{Name: className, Filename: gFunction.File, Methods: []Method{}, Lines: []Line{}}
+				class = &Class{Name: className, Filename: gFunction.File, Methods: []Method{}, Lines: []Line{}, LineCount: 0, LineHits: 0}
```
Comment (AlekSi) [inline gocov-xml.go]: Go always initializes everything with zero values, no need to be explicit about it.

File: gocov-xml.go [hunk 9]
```diff
@@ -109,30 +122,43 @@ func main() {
...
 
-			class.Methods = append(class.Methods, Method{Name: methodName, Lines: lines})
+			class.Methods = append(class.Methods, Method{Name: methodName, Lines: lines, LineRate: line_rate})
+			class.LineCount += int64(len(gFunction.Statements))
+			class.LineHits += int64(func_hits)
 		}
 
 		// fill package with "classes"
-		p := Package{Name: gPackage.Name, Classes: []Class{}}
+		p := Package{Name: gPackage.Name, Classes: []Class{}, LineCount: 0, LineHits: 0}
```
Comment (AlekSi) [inline gocov-xml.go]: Drop `0`.

File: gocov-xml.go (hunk 8 above)
Comment (yaoyaozong) [reply gocov-xml.go]: not find a way for how to avoid 0 in structs field:value init. 
Sorry not too much experience on Go...

**Comment by yaoyaozong** [comment]
Done. Made the changes according to the comments in PR except for the field:value initialize on structs

**Comment by yaoyaozong** [comment]
Done! Thanks for the comments. Learned a new way to init in Go! ^_^

**Comment by AlekSi** [comment]
Merged, thank you!
